### DISCUSSION TITLE
Drawing custom widgets using GLSL Fragment Shaders

### DISCUSSION THREAD

**Discussion by Aligorith** [emilk/egui#620] [Q&A]
I'd like to be able to embed several custom widgets, each one drawn using a different GLSL fragment shader, within a UI layout generated with egui  (e.g. one large viewport + one smaller preview).  What's the best way of doing something like this?

Currently, it looks like the only way is to do the following, but I was wondering if there's a way that doesn't require copying the pixel data back and forth like this (i.e. GPU -> RAM -> GPU):
1) Render the shader to an appropriately-sized offscreen buffer using glium or something
2) Copy that buffer's pixel data into a egui Texture
3) Assign that texture to an egui::Image widget

**Comment by emilk** [comment]
You can render it to a texture and show that directly using `TextureId::User` - there is no need to copy it to RAM and into a `egui::Texture`.
